File: hi_jim_android/app/src/main/java/com/jim/hi_jim/presentation/viewmodel/mukjjippa/MukjjippaGameLobbyViewModel.kt
package com.jim.hi_jim.presentation.viewmodel.mukjjippa

import android.util.Log
import androidx.lifecycle.ViewModel
import androidx.lifecycle.viewModelScope
import com.jim.hi_jim.data.repository.FirebaseGameRepository
import com.jim.hi_jim.presentation.constants.UserConstants
import com.jim.hi_jim.shared.model.GameRequest
import com.jim.hi_jim.shared.model.GameRequestStatus
import com.jim.hi_jim.shared.model.GameType
import kotlinx.coroutines.flow.*
import kotlinx.coroutines.launch

class MukjjippaGameLobbyViewModel : ViewModel() {

    private val repository = FirebaseGameRepository()
    private val currentUserId = UserConstants.CURRENT_USER_ID
    private val otherUserId = if (currentUserId == UserConstants.USER_1)
        UserConstants.USER_2 else UserConstants.USER_1

    // 받은 묵찌빠 게임 요청 목록
    val receivedRequests: StateFlow<List<GameRequest>> = repository
        .observeGameRequestsByType(currentUserId, GameType.MUKJJIPPA)
        .stateIn(
            scope = viewModelScope,
            started = SharingStarted.WhileSubscribed(5000),
            initialValue = emptyList()
        )

    // 보낸 요청 ID
    private val _sentRequestId = MutableStateFlow<String?>(null)

    // 보낸 요청의 전체 정보 (상태 + gameId 포함)
    private val sentRequest: StateFlow<GameRequest?> = _sentRequestId
        .flatMapLatest { requestId ->
            if (requestId != null) {
                repository.observeSentRequest(currentUserId, requestId)
            } else {
                flowOf(null)
            }
        }
        .stateIn(
            scope = viewModelScope,
            started = SharingStarted.Eagerly,
            initialValue = null
        )

    // 보낸 요청 상태 (UI 표시용)
    val sentRequestStatus: StateFlow<GameRequestStatus?> = sentRequest
        .map { it?.status }
        .stateIn(
            scope = viewModelScope,
            started = SharingStarted.WhileSubscribed(5000),
            initialValue = null
        )

    // 생성된 게임 ID
    private val _gameId = MutableStateFlow<String?>(null)
    val gameId: StateFlow<String?> = _gameId.asStateFlow()

    companion object {
        private const val TAG = "MukjjippaGameLobbyVM"
    }

    // 묵찌빠 게임 요청 보내기
    fun sendGameRequest() {
        viewModelScope.launch {
            val result = repository.sendGameRequest(currentUserId, otherUserId, GameType.MUKJJIPPA)
            result.onSuccess { requestId ->
                _sentRequestId.value = requestId
                Log.d(TAG, "Mukjjippa request sent: $requestId")
            }.onFailure { error ->
                Log.e(TAG, "Failed to send mukjjippa request", error)
            }
        }
    }

    // 게임 요청 수락
    fun acceptGameRequest(requestId: String) {
        viewModelScope.launch {
            val result = repository.respondToGameRequest(currentUserId, requestId, accept = true)
            result.onSuccess { gameId ->
                if (gameId != null) {
                    _gameId.value = gameId
                    Log.d(TAG, "Mukjjippa game created: $gameId")
                }
            }.onFailure { error ->
                Log.e(TAG, "Failed to accept mukjjippa request", error)
            }
        }
    }

    // 게임 요청 거절
    fun rejectGameRequest(requestId: String) {
        viewModelScope.launch {
            repository.respondToGameRequest(currentUserId, requestId, accept = false)
        }
    }

    // 요청 취소
    fun cancelGameRequest() {
        viewModelScope.launch {
            val requestId = _sentRequestId.value
            if (requestId != null) {
                repository.respondToGameRequest(currentUserId, requestId, accept = false)
                _sentRequestId.value = null
                Log.d(TAG, "Mukjjippa request cancelled: $requestId")
            }
        }
    }

    // 보낸 요청이 수락되었을 때 게임 ID 확인
    init {
        viewModelScope.launch {
            sentRequest.collect { request ->
                Log.d(TAG, "Sent mukjjippa request update: status=${request?.status}, gameId=${request?.gameId}")

                when {
                    // 요청이 수락되고 gameId가 있으면 게임 시작
                    request?.status == GameRequestStatus.ACCEPTED && request.gameId != null -> {
                        _gameId.value = request.gameId
                        _sentRequestId.value = null
                        Log.d(TAG, "✅ Mukjjippa game started from request sender: ${request.gameId}")
                    }
                    // 요청이 삭제된 경우 (거절/취소)
                    request == null && _sentRequestId.value != null -> {
                        _sentRequestId.value = null
                        Log.d(TAG, "🔴 Mukjjippa request rejected or cancelled")
                    }
                }
            }
        }
    }
}

File: hi_jim_ios/hi_jim Watch App/ViewModels/MukjjippaGameLobbyViewModel.swift
//
//  MukjjippaGameLobbyViewModel.swift
//  hi_jim Watch App
//
//  묵찌빠 게임 로비 상태 관리
//

import Foundation
import Combine

@MainActor
class MukjjippaGameLobbyViewModel: ObservableObject {
    @Published var receivedRequests: [GameRequest] = []
    @Published var sentRequestStatus: GameRequestStatus?
    @Published var currentGameId: String?
    @Published var isLoading = false
    @Published var errorMessage: String?

    private let repository = FirebaseGameRepository()
    private var cancellables = Set<AnyCancellable>()
    private var currentRequestId: String?

    let currentUserId = UserConstants.CURRENT_USER_ID
    let opponentUserId = UserConstants.opponentUserId

    // 초기화 시 받은 요청 감지 시작
    init() {
        observeReceivedRequests()
    }

    // MARK: - Observe Requests

    // 받은 게임 요청 감지 (묵찌빠만)
    private func observeReceivedRequests() {
        repository.observeGameRequestsByType(userId: currentUserId, gameType: .MUKJJIPPA)
            .receive(on: DispatchQueue.main)
            .sink(
                receiveCompletion: { [weak self] completion in
                    if case .failure(let error) = completion {
                        self?.errorMessage = "요청 감지 오류: \(error.localizedDescription)"
                    }
                },
                receiveValue: { [weak self] requests in
                    self?.receivedRequests = requests
                }
            )
            .store(in: &cancellables)
    }

    // 보낸 요청 상태 감지
    private func observeSentRequestStatus(requestId: String) {
        repository.observeSentRequest(fromUserId: currentUserId, requestId: requestId)
            .receive(on: DispatchQueue.main)
            .sink(
                receiveCompletion: { [weak self] completion in
                    if case .failure(let error) = completion {
                        self?.errorMessage = "요청 상태 감지 오류: \(error.localizedDescription)"
                    }
                },
                receiveValue: { [weak self] request in
                    guard let self = self else { return }

                    if let request = request {
                        print("🟡 [MukjjippaLobby] Sent request update: status=\(request.status), gameId=\(request.gameId ?? "nil")")
                        self.sentRequestStatus = request.status

                        // 수락되면 게임 ID 저장하고 게임 화면으로 이동
                        if request.status == .accepted {
                            if let gameId = request.gameId {
                                print("✅ [MukjjippaLobby] Game started from request sender: \(gameId)")
                                self.currentGameId = gameId
                                self.sentRequestStatus = nil
                                self.currentRequestId = nil
                            } else {
                                print("🔴 [MukjjippaLobby] Request accepted but gameId is nil!")
                            }
                        }
                    } else {
                        print("🟡 [MukjjippaLobby] Request deleted (rejected/cancelled)")
                        self.sentRequestStatus = nil
                        self.currentRequestId = nil
                    }
                }
            )
            .store(in: &cancellables)
    }

    // MARK: - Actions

    // 게임 요청 보내기
    func sendGameRequest() {
        print("🟡 [MukjjippaLobby] Sending game request from \(currentUserId) to \(opponentUserId)")
        isLoading = true
        errorMessage = nil

        Task {
            do {
                let requestId = try await repository.sendGameRequest(
                    fromUserId: currentUserId,
                    toUserId: opponentUserId,
                    gameType: .MUKJJIPPA
                )

                print("✅ [MukjjippaLobby] Request sent with ID: \(requestId)")
                currentRequestId = requestId
                sentRequestStatus = .pending
                observeSentRequestStatus(requestId: requestId)
                isLoading = false

            } catch {
                print("🔴 [MukjjippaLobby] Send request failed: \(error.localizedDescription)")
                errorMessage = "요청 전송 실패: \(error.localizedDescription)"
                isLoading = false
            }
        }
    }

    // 보낸 요청 취소
    func cancelSentRequest() {
        guard let requestId = currentRequestId else { return }

        isLoading = true

        Task {
            do {
                _ = try await repository.respondToGameRequest(
                    userId: currentUserId,
                    requestId: requestId,
                    accept: false
                )

                sentRequestStatus = nil
                currentRequestId = nil
                isLoading = false

            } catch {
                errorMessage = "요청 취소 실패: \(error.localizedDescription)"
                isLoading = false
            }
        }
    }

    // 받은 요청 수락
    func acceptRequest(_ request: GameRequest) {
        print("🟡 [MukjjippaLobby] Accepting request: \(request.requestId)")
        isLoading = true
        errorMessage = nil

        Task {
            do {
                if let gameId = try await repository.respondToGameRequest(
                    userId: currentUserId,
                    requestId: request.requestId,
                    accept: true
                ) {
                    print("✅ [MukjjippaLobby] Game created with ID: \(gameId)")
                    currentGameId = gameId
                } else {
                    print("🔴 [MukjjippaLobby] respondToGameRequest returned nil gameId")
                }
                isLoading = false

            } catch {
                print("🔴 [MukjjippaLobby] Accept request failed: \(error.localizedDescription)")
                errorMessage = "요청 수락 실패: \(error.localizedDescription)"
                isLoading = false
            }
        }
    }

    // 받은 요청 거절
    func rejectRequest(_ request: GameRequest) {
        Task {
            do {
                _ = try await repository.respondToGameRequest(
                    userId: currentUserId,
                    requestId: request.requestId,
                    accept: false
                )
            } catch {
                errorMessage = "요청 거절 실패: \(error.localizedDescription)"
            }
        }
    }

    // 게임에서 돌아왔을 때 게임 ID 초기화
    func resetGame() {
        currentGameId = nil
    }
}
